lib/src/dag_walk.rs

```diff
@@ -93,6 +93,151 @@ where
+    emitted: HashSet<ID>,
+}
+
+impl<T: Ord, ID: Hash + Eq + Clone> TopoOrderReverseLazyInner<T, ID> {
+    fn new(start: Vec<T>) -> Self {
+        TopoOrderReverseLazyInner {
+            start,
+            result: Vec::new(),
+            emitted: HashSet::new(),
+        }
+    }
+
+    fn next<NI: IntoIterator<Item = T>>(
+        &mut self,
+        id_fn: impl Fn(&T) -> ID,
+        mut neighbors_fn: impl FnMut(&T) -> NI,
+    ) -> Option<T> {
+        if let Some(node) = self.result.pop() {
+            return Some(node);
+        }
+
+        // Fast path for linear DAG
+        if self.start.len() <= 1 {
+            let node = self.start.pop()?;
+            self.start.extend(neighbors_fn(&node));
+            assert!(self.emitted.insert(id_fn(&node)), "graph has cycle");
+            return Some(node);
+        }
+
+        // Extract graph nodes based on T's order, and sort them by using ids
+        // (because we wouldn't want to clone T itself)
+        let start_ids = self.start.iter().map(&id_fn).collect_vec();
+        let (mut node_map, neighbor_ids_map, remainder) =
+            look_ahead_sub_graph(mem::take(&mut self.start), &id_fn, &mut neighbors_fn);
+        self.start = remainder;
+        let sorted_ids = topo_order_forward(&start_ids, |id| *id, |id| &neighbor_ids_map[id]);
+        self.result.reserve(sorted_ids.len());
+        for id in sorted_ids {
+            let (id, node) = node_map.remove_entry(id).unwrap();
+            assert!(self.emitted.insert(id), "graph has cycle");
+            self.result.push(node);
+        }
+        self.result.pop()
+    }
+}
+
+/// Splits DAG at single fork point, and extracts branchy part as sub graph.
+///
+/// ```text
+///  o | C
+///  | o B
+///  |/ <---- split here (A->B or A->C would create cycle)
+///  o A
+/// ```
+///
+/// If a branch reached to root (empty neighbors), the graph can't be split
+/// anymore because the other branch may be connected to a descendant of
+/// the rooted branch.
+///
+/// ```text
+///  o | C
+///  | o B
+///  |  <---- can't split here (there may be edge A->B)
+///  o A
+/// ```
+///
+/// We assume the graph is (mostly) topologically ordered by `T: Ord`.
+#[allow(clippy::type_complexity)]
+fn look_ahead_sub_graph<T, ID, NI>(
+    start: Vec<T>,
+    id_fn: impl Fn(&T) -> ID,
+    mut neighbors_fn: impl FnMut(&T) -> NI,
+) -> (HashMap<ID, T>, HashMap<ID, Vec<ID>>, Vec<T>)
+where
+    T: Ord,
+    ID: Hash + Eq + Clone,
+    NI: IntoIterator<Item = T>,
+{
+    let mut queue: BinaryHeap<T> = start.into();
+    // Build separate node/neighbors maps since lifetime is different at caller
+    let mut node_map: HashMap<ID, T> = HashMap::new();
+    let mut neighbor_ids_map: HashMap<ID, Vec<ID>> = HashMap::new();
+    let mut has_reached_root = false;
+    while queue.len() > 1 || node_map.is_empty() || has_reached_root {
+        let node = if let Some(node) = queue.pop() {
+            node
+        } else {
+            break;
+        };
```

Comment:
nit: Would it be simpler to do `while let Some(node) = queue.pop() {` here and `if queue.len() <= 1 && !has_reached_root {break; }` at the end of the block? Just a thought; I'm fine with either.

Reply:
Actually I tried that before, but the problem is I (ab)use `node_map.contains_key(&node_id) { continue; }` to deduplicate queued nodes, so the break condition would have to be added there.